src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <exception>
+#include <map>
+#include <set>
+#include <pthread.h>
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+#include <cudf_jni_apis.hpp>
+
+namespace {
+
+constexpr char const *RMM_EXCEPTION_CLASS = "ai/rapids/cudf/RmmException";
+
+class thread_state {
+public:
+    bool retry_oom_injected = false;
+    bool split_and_retry_oom_injected = false;
+    long task_id = -1;
+};
+
+template<typename T>
+class rollback {
```

Comment:
I would prefer `std::function<void()>` as the type instead of the template (but I do not know if other C++ experts agree). It seems we always want this to be a void function that takes no arguments.